apps/query-ui/queryui/util.py

```diff
@@ -66,105 +80,196 @@ def docset_to_string(docset: DocSet) -> str:
+    return bucket, key
+
+
+class PDFPreview:
+    """Display a preview of the given PDF file."""
+
+    def __init__(self, path: str):
+        self.path = path
+
+    def show(self):
+        if self.path.startswith("s3://"):
+            bucket, key = parse_s3_path(self.path)
+            s3 = boto3.client("s3")
+            response = s3.get_object(Bucket=bucket, Key=key)
+            content = response["Body"].read()
+        elif self.path.startswith("http"):
+            content = requests.get(self.path, timeout=30).content
+        else:
+            st.write(f"Unknown path format: {self.path}")
+            return
+
+        st.text(self.path)
+        encoded = base64.b64encode(content).decode("utf-8")
+        pdf_display = (
+            f'<iframe src="data:application/pdf;base64,{encoded}" '
+            + 'width="600" height="800" type="application/pdf"></iframe>'
         )
-    for node in plan.nodes.values():
-        if node.dependencies:
-            for dep in node.dependencies:
-                edges.append(Edge(source=dep.node_id, target=node.node_id, color="#ffffff"))
-
-    config = Config(
-        width=700,
-        height=500,
-        directed=True,
-        physics=False,
-        hierarchical=True,
-        direction="UD",
-    )
-    agraph(nodes=nodes, edges=edges, config=config)
+        st.markdown(pdf_display, unsafe_allow_html=True)
+
+
+@st.dialog("Document preview", width="large")
+def show_pdf_preview(path: str):
+    """Display a preview of the given document in a dialog."""
+    PDFPreview(path).show()
+
+
+class QueryNodeTrace:
+    """Helper class to read and display the trace of a single query node."""
+
+    # The order here is chosen to ensure that the most important columns are shown first.
+    # The logic below ensure that additional columns are also shown, and that empty columns
+    # are omitted.
+    COLUMNS = [
+        "properties.path",
+        "properties.key",
+        "properties.count",
+        "doc_id",
+        "parent_id",
+        "text_representation",
+    ]
+
+    def __init__(self, trace_dir: str, node_id: str):
+        self.trace_dir = trace_dir
+        self.node_id = node_id
+        self.df = None
+        self.readdata()
+
+    def readfile(self, f):
+        """Read the given trace file."""
+        with open(f, "rb") as file:
+            try:
+                doc = pickle.load(file)
+            except EOFError:
+                return None
+
+            # For now, skip over MetadataDocuments.
+            if "metadata" in doc.keys():
+                return None
+
+            # Flatten properties.
+            if "properties" in doc:
+                for prop in doc["properties"]:
+                    if isinstance(doc["properties"][prop], dict):
+                        for nested_property in doc["properties"][prop]:
+                            doc[".".join(["properties", prop, nested_property])] = doc["properties"][prop][
```

Comment:
for some cases I needed to convert the value to a string as well (multiple inferred types for the same key). So I'd do a str(doc["properties"][prop][next_property]) since this is for displaying only

Reply:
So, the problem with that is that casting to str doesn't let streamlit do its magic with rendering, e.g., lists as individual items. Is the issue only for multiple types with the same key? I'm a little confused why we would end up with different types for the same key though.